mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala

```diff
@@ -153,102 +48,75 @@ private[feature] trait ChiSqSelectorParams extends Params
   @Since("1.6.0")
-  def setLabelCol(value: String): this.type = set(labelCol, value)
-
-  @Since("2.0.0")
-  override def fit(dataset: Dataset[_]): ChiSqSelectorModel = {
-    transformSchema(dataset.schema, logging = true)
-    dataset.select(col($(labelCol)).cast(DoubleType), col($(featuresCol))).rdd.map {
-      case Row(label: Double, features: Vector) =>
-        LabeledPoint(label, features)
-    }
+  override def setLabelCol(value: String): this.type = super.setLabelCol(value)
 
-    val testResult = ChiSquareTest.testChiSquare(dataset, getFeaturesCol, getLabelCol)
-      .zipWithIndex
-    val features = $(selectorType) match {
-      case "numTopFeatures" =>
-        testResult
-          .sortBy { case (res, _) => res.pValue }
-          .take(getNumTopFeatures)
-      case "percentile" =>
-        testResult
-          .sortBy { case (res, _) => res.pValue }
-          .take((testResult.length * getPercentile).toInt)
-      case "fpr" =>
-        testResult
-          .filter { case (res, _) => res.pValue < getFpr }
-      case "fdr" =>
-        // This uses the Benjamini-Hochberg procedure.
-        // https://en.wikipedia.org/wiki/False_discovery_rate#Benjamini.E2.80.93Hochberg_procedure
-        val tempRes = testResult
-          .sortBy { case (res, _) => res.pValue }
-        val selected = tempRes
-          .zipWithIndex
-          .filter { case ((res, _), index) =>
-            res.pValue <= getFdr * (index + 1) / testResult.length }
-        if (selected.isEmpty) {
-          Array.empty[(SelectionTestResult, Int)]
-        } else {
-          val maxIndex = selected.map(_._2).max
-          tempRes.take(maxIndex + 1)
-        }
-      case "fwe" =>
-        testResult
-          .filter { case (res, _) => res.pValue < getFwe / testResult.length }
-      case errorType =>
-        throw new IllegalStateException(s"Unknown Selector Type: $errorType")
-    }
-    val indices = features.map { case (_, index) => index }
-    copyValues(new ChiSqSelectorModel(uid, indices.sorted)
-      .setParent(this))
+  /**
+   * get the SelectionTestResult for every feature against the label
+   */
+  @Since("3.1.0")
+  protected[this] override def getSelectionTestResult(dataset: Dataset[_]):
+  Array[SelectionTestResult] = {
+    SelectionTest.chiSquareTest(dataset, getFeaturesCol, getLabelCol)
   }
 
-  @Since("1.6.0")
-  override def transformSchema(schema: StructType): StructType = {
-    SchemaUtils.checkColumnType(schema, $(featuresCol), new VectorUDT)
-    SchemaUtils.checkNumericType(schema, $(labelCol))
-    SchemaUtils.appendColumn(schema, $(outputCol), new VectorUDT)
+  /**
+   * Create a new instance of concrete SelectorModel.
+   * @param indices The indices of the selected features
+   * @param pValues The pValues of the selected features
+   * @param statistics The chi square statistic of the selected features
+   * @return A new SelectorModel instance
+   */
+  @Since("3.1.0")
+  protected[this] def createSelectorModel(
```

Comment:
ditto

Reply:
this method is just a constructor, so I think we do not need it.

Reply:
I am not sure if I can delete this method. I have problem doing the following in the end of ```Selector.fit```
```
copyValues(new SelectorModel(uid, indices, pValues, statistic).setParent(this))
```